Remove many usages of Type

li-roy opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21964 Remove Type dispatch
* #21963 Stop using Type in Python bindings
* **#21941 Remove many usages of Type**
* #21940 Remove getDeviceFromPtr and allocator from Type
* #21923 Move UnsafeTensorFromTH and UnsafeStorageFromTH off Type
* #21320 [BC-BREAKING] Replace Type dispatch with ATenDispatch

Removes many usages of type, in preparation for deletion:
- Use of Type in python binding for constructors. replaced with TensorTypeId
- Use of Type to save python default tensor. replaced with TensorTypeId

Differential Revision: [D15893331](https://our.internmc.facebook.com/intern/diff/D15893331)

Commit:
Remove many usages of Type

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head

ezyang commented on torch/csrc/autograd/VariableTypeManual.cpp:
```diff
@@ -117,25 +117,24 @@ at::TypeExtendedInterface* VariableType::getVariableTypeFromBaseType(const at::T
   for (auto p : backends) {
-    auto baseType = context.getNonVariableTypeRaw(static_cast<Backend>(p), ScalarType::Undefined);
-    if (baseType) {
-      res.emplace_back(VariableType::getVariableTypeFromBaseType(*baseType));
```
Weird, how come we didn't previously loop?

ezyang commented on torch/csrc/tensor/python_tensor.cpp:
```diff
@@ -242,6 +238,32 @@ static THPObjectPtr get_tensor_dict() {
 
 static std::vector<PyTensorType> tensor_types;
 
+void set_default_tensor_type(PyTensorType* type) {
```
This got moved, are there any substantive changes?

ezyang commented on torch/csrc/utils/tensor_types.cpp:
```diff
@@ -62,8 +68,7 @@ std::pair<at::Type*, at::ScalarType> type_from_string(const std::string& str) {
         for (int s = 0; s < static_cast<int>(ScalarType::NumOptions); s++) {
-          cpu_map.emplace(type_to_string(*type, static_cast<ScalarType>(s)),
-                          std::make_pair(type, static_cast<ScalarType>(s)));
+          cpu_map.emplace(type_to_string(*type), type);
```
Good!

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head

gchanan commented on torch/csrc/Module.cpp:
```diff
@@ -452,7 +452,7 @@ PyObject *THPModule_setFlushDenormal(PyObject *_unused, PyObject *arg) {
 PyObject *THPModule_getDefaultDtype(PyObject *_unused, PyObject *arg) {
   HANDLE_TH_ERRORS
-  auto scalar_type = torch::tensors::get_default_scalar_type();
+  auto scalar_type = at::typeMetaToScalarType(torch::tensors::get_default_tensor_options().dtype());
```
I don't think we should do this.

There is no such thing as default tensor options.  There's a default dtype, and a default device type for constructors, but that's it.  Using tensor options for this implies there are dimensions of the default which are absolutely not applicable: currently layout, optionality, etc.  But soon there will also be dimension names and that doesn't make sense in this context.

The right way to think about this, I think, is that TensorOptions are a mechanism for C++ users to specify kwarg-like options to our factory function.  Every other use of them has been a mistake (`Type.options()`, hacking up our parsing in multiple places to pass TensorOptions instead of just the actual parameters, etc.)

li-roy replied:
Okay, I'll replace this with get_default_tensor_type_id, which is better than get_default_tensor_type because we are still having a default backend, but we just don't have scalar_type info passed along with it.

li-roy replied:
In the future we should probably have a default device type instead.

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head

li-roy commented on tools/autograd/templates/VariableType.h:
```diff
@@ -56,8 +56,8 @@ struct TORCH_API VariableType final : public at::TypeDefault {
   static bool isVariableType(const at::Type& type);
-  static std::vector<at::Type*> allCUDATypes();
-  static std::vector<at::Type*> allCPUTypes();
+  static std::vector<at::DeprecatedTypeProperties*> allCUDATypes();
```
It's likely we can replace with TensorTypeId in the future.

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head

gchanan commented on torch/csrc/utils/tensor_types.cpp:
```diff
@@ -14,46 +15,46 @@ using namespace at;
     // torch.cuda. is prefix of str
     std::call_once(cuda_once, []() {
       for (auto type : autograd::VariableType::allCUDATypes()) {
         for (int s = 0; s < static_cast<int>(ScalarType::NumOptions); s++) {
```
wait, do you still need to loop over ScalarTypes here?

li-roy replied:
Yeah, we still need to return Backend, ScalarType pair.

gchanan replied:
but you are never using `s`?

li-roy replied:
Oh I see what you mean

gchanan commented on torch/csrc/utils/tensor_types.cpp:
```diff
@@ -62,8 +63,7 @@ std::pair<at::Type*, at::ScalarType> type_from_string(const std::string& str) {
     std::call_once(cpu_once, []() {
       for (auto type : autograd::VariableType::allCPUTypes()) {
         for (int s = 0; s < static_cast<int>(ScalarType::NumOptions); s++) {
```
same here.

gchanan commented on test/common_nn.py:
```diff
@@ -3189,7 +3189,7 @@ def test_noncontig(self, test_case, module, input):
         with freeze_rng_state():
             output = test_case._forward(module, input)
-            grad_output = output.new(output.shape).normal_()
+            grad_output = output.new(output.shape, device=output.device).normal_()
```
is this necessary?  If so, that's broken.

gchanan commented on torch/csrc/tensor/python_tensor.cpp:
```diff
@@ -327,54 +348,30 @@ void py_set_default_tensor_type(PyObject* obj) {
-    throw unavailable_type(*type);
-  }
-  set_default_tensor_type(*aten_type, scalar_type);
+  set_default_tensor_type(type);
```
I don't know if this is due to this specific change or somewhere else, but this seems worse:

Before you patch (without CUDA):
```
torch.set_default_tensor_type(torch.cuda.FloatTensor)
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-2-876b829d82b3> in <module>()
----> 1 torch.set_default_tensor_type(torch.cuda.FloatTensor)

/data/users/gchanan/_pytorch11/torch/__init__.py in set_default_tensor_type(t)
    151     if isinstance(t, _string_classes):
    152         t = _import_dotted_name(t)
--> 153     _C._set_default_tensor_type(t)
    154
    155

/data/users/gchanan/_pytorch11/torch/cuda/__init__.py in _lazy_init()
    176         raise RuntimeError(
    177             "Cannot re-initialize CUDA in forked subprocess. " + msg)
--> 178     _check_driver()
    179     torch._C._cuda_init()
    180     _cudart = _load_cudart()

/data/users/gchanan/_pytorch11/torch/cuda/__init__.py in _check_driver()
     90 def _check_driver():
     91     if not hasattr(torch._C, '_cuda_isDriverSufficient'):
---> 92         raise AssertionError("Torch not compiled with CUDA enabled")
     93     if not torch._C._cuda_isDriverSufficient():
     94         if torch._C._cuda_getDriverVersion() == 0:

AssertionError: Torch not compiled with CUDA enabled
```

with your patch:
```
In [4]: torch.set_default_tensor_type(torch.cuda.FloatTensor)

In [5]: torch.randn(2,3)
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-5-1dc2fcf8de67> in <module>()
----> 1 torch.randn(2,3)

/data/users/gchanan/_pytorch8/torch/cuda/__init__.py in _lazy_init()
    176         raise RuntimeError(
    177             "Cannot re-initialize CUDA in forked subprocess. " + msg)
--> 178     _check_driver()
    179     torch._C._cuda_init()
    180     _cudart = _load_cudart()

/data/users/gchanan/_pytorch8/torch/cuda/__init__.py in _check_driver()
     90 def _check_driver():
     91     if not hasattr(torch._C, '_cuda_isDriverSufficient'):
---> 92         raise AssertionError("Torch not compiled with CUDA enabled")
     93     if not torch._C._cuda_isDriverSufficient():
     94         if torch._C._cuda_getDriverVersion() == 0:

AssertionError: Torch not compiled with CUDA enabled
```

li-roy replied:
fixed and added a test for it.

gchanan commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -53,39 +60,51 @@ void maybe_initialize_cuda(const Device device) {
+  maybe_initialize_cuda(type_id);
   AutoNoGIL no_gil;
-  return torch::zeros(sizes, type.options(scalar_type, std::move(device)));
+  if (device.has_value()) {
```
for all of these that condition based on whether the device exists...can we just build that into your `options` function?

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head

li-roy commented on aten/src/ATen/detail/CUDAHooksInterface.h:
```diff
@@ -23,7 +23,8 @@ namespace at {
   "PyTorch splits its backend into two shared libraries: a CPU library "
   "and a CUDA library; this error has occurred because you are trying "
-  "to use some CUDA functionality, but the CUDA library has not been "
+  "to use some CUDA functionality, but the CUDA library does not exist."
```
This change isn't needed anymore, I'll revert.

Initially, this PR made it so that this was being hit in some cases when pytorch was being built without CUDA, and I wanted to include that possibility in the error message. But the most recent changes make it so it won't be hit anymore.

li-roy commented on test/test_torch.py:
```diff
@@ -11734,7 +11734,7 @@ def test_cuda_not_built(self):
         self.assertRaisesRegex(AssertionError, msg, lambda: torch.tensor([1], device="cuda"))
         self.assertRaisesRegex(AssertionError, msg, lambda: torch.tensor([1]).cuda())
-        self.assertRaisesRegex(AssertionError, msg, lambda: torch.cuda.FloatTensor())
+        self.assertRaisesRegex(RuntimeError, msg, lambda: torch.cuda.FloatTensor())
```
It's hitting a different error now. After most recent changes, it hits an error saying that cuda.FloatTensor isn't available.

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head

li-roy commented on torch/csrc/tensor/python_tensor.cpp:
```diff
@@ -67,12 +66,7 @@ static TypeError unavailable_type(const PyTensorType& type) {
-  }
-  auto scalar_type = static_cast<ScalarType>(tensor_type.scalar_type);
-  return THPVariable_Wrap(torch::utils::legacy_tensor_ctor(*aten_type, scalar_type, args, kwargs));
+  return THPVariable_Wrap(torch::utils::legacy_tensor_ctor(tensor_type.get_type_id(), tensor_type.get_scalar_type(), args, kwargs));
```
This occurs before my change too, right? If so, I don't have MKLDNN built, but my guess is that `new` calls `legacy_new_from_sequence` which calls `internal_new_from_data`, which uses `to`, so it doesn't support layouts.

gchanan replied:
yes, it does.  I'll file an issue.

gchanan replied:
https://github.com/pytorch/pytorch/issues/22158

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head

Commit:
Update on "Remove many usages of Type"

Remove many usages of Type

gh-metadata: pytorch pytorch 21941 gh/li-roy/30/head
